src/OpenTelemetry.Instrumentation.EventCounters/EventCounterListener.cs

```diff
@@ -0,0 +1,199 @@
+            }
+        }
+
+        private void ExtractAndPostMetric(string eventSourceName, IDictionary<string, object> eventPayload)
+        {
+            try
+            {
+                bool calculateRate = false;
+                double actualValue = 0;
+
+                string counterName = string.Empty;
+                string counterDisplayName = string.Empty;
+                InstrumentType instrumentType = InstrumentType.ObservableGauge;
+
+                foreach (KeyValuePair<string, object> payload in eventPayload)
+                {
+                    var key = payload.Key;
+                    if (key.Equals("Name", StringComparison.OrdinalIgnoreCase))
+                    {
+                        counterName = payload.Value.ToString();
+                    }
+                    else
+                    if (key.Equals("DisplayName", StringComparison.OrdinalIgnoreCase))
+                    {
+                        counterDisplayName = payload.Value.ToString();
+                    }
+                    else if (key.Equals("Mean", StringComparison.OrdinalIgnoreCase))
+                    {
+                        instrumentType = InstrumentType.ObservableGauge;
+                        actualValue = Convert.ToDouble(payload.Value);
+                    }
+                    else if (key.Equals("Increment", StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Increment indicates we have to calculate rate.
+                        instrumentType = InstrumentType.ObservableCounter;
+                        calculateRate = true;
+                        actualValue = Convert.ToDouble(payload.Value);
+                    }
+                }
+
+                this.RecordMetric(eventSourceName, counterName, counterDisplayName, instrumentType, actualValue, calculateRate);
+            }
+            catch (Exception ex)
+            {
+                EventCountersInstrumentationEventSource.Log.EventCountersInstrumentationWarning("ExtractMetric", ex.Message);
+            }
+        }
+
+        private void RecordMetric(string eventSourceName, string counterName, string displayName, InstrumentType instrumentType, double value, bool calculateRate)
+        {
+            var metricKey = new MetricKey(eventSourceName, counterName);
+            var description = string.IsNullOrEmpty(displayName) ? counterName : displayName;
+            this.lastValue[metricKey] = calculateRate ? value / this.options.RefreshIntervalSecs : value;
+            switch (instrumentType)
+            {
+                case InstrumentType.ObservableCounter:
+                    if (!this.metricInstruments.ContainsKey(metricKey))
+                    {
+                        this.metricInstruments[metricKey] = this.meter.CreateObservableCounter(counterName, () => this.ObserveDouble(metricKey), description: description);
```

Comment:
Use the `TryAdd` method of the `ConcurrentDictionary` to add instruments for thread-safe implementation.